File: Java/Q203RemoveLinkedListElements.java
/*
@b-knd (jingru) on 31 July 2022 16:55:00
*/

class Solution {
    public ListNode removeElements(ListNode head, int val) {
        ListNode dummyhead = new ListNode();
        ListNode prev = dummyhead;
        ListNode curr = head;
        dummyhead.next = curr;
        
        /*
        - if current value is the same as value to be removed, set prev's next to current next (remove current node), and advace to next node for current
        - else, advance both previous and current node
        */
        while(curr != null){
            if(curr.val == val){
                prev.next = curr.next;
            } else{
                prev = prev.next;
            }
            curr = curr.next;
        }
        
        return dummyhead.next;
    }
}

//Runtime: 0 ms, faster than 100.00% of Java online submissions for Remove Linked List Elements.
//Memory Usage: 43.3 MB, less than 89.50% of Java online submissions for Remove Linked List Elements.

File: Python/Q203RemoveLinkedListElements.py
# @b-knd (jingru) on 09 August 2022 11:25:00

# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def removeElements(self, head: Optional[ListNode], val: int) -> Optional[ListNode]:
        dummyHead = ListNode(0)
        curr = head
        dummyHead.next = curr
        prev = dummyHead
        
        while curr != None:
            while curr != None and curr.val == val:
                curr = curr.next
            prev.next = curr
            prev = curr
            if curr != None:
                curr = curr.next
        return dummyHead.next
    
#Runtime: 72 ms, faster than 92.62% of Python3 online submissions for Remove Linked List Elements.
#Memory Usage: 17.8 MB, less than 81.81% of Python3 online submissions for Remove Linked List Elements.
